src/binding/c/mpit_api.txt

```diff
@@ -0,0 +1,515 @@
+{
+    name2index_hash_t *hash_entry;
+
+    /* Do hash lookup by the name */
+    HASH_FIND_STR(cat_hash, name, hash_entry);
+    if (hash_entry != NULL) {
+        *cat_index = hash_entry->idx;
+    } else {
+        mpi_errno = MPI_T_ERR_INVALID_NAME;
+        goto fn_fail;
+    }
+}
+
+MPI_T_category_get_info:
+    .desc: Get the information about a category
+    .skip: validate-ANY
+    .error: MPI_T_ERR_INVALID_INDEX
+{ -- error_check --
+    MPIT_ERRTEST_CAT_INDEX(cat_index);
+}
+{
+    cat_table_entry_t *cat;
+    cat = (cat_table_entry_t *) utarray_eltptr(cat_table, cat_index);
+    MPIR_T_strncpy(name, cat->name, name_len);
+    MPIR_T_strncpy(desc, cat->desc, desc_len);
+
+    if (num_cvars != NULL)
+        *num_cvars = utarray_len(cat->cvar_indices);
+
+    if (num_pvars != NULL)
+        *num_pvars = utarray_len(cat->pvar_indices);
+
+    if (num_categories != NULL)
+        *num_categories = utarray_len(cat->subcat_indices);
+}
+
+MPI_T_category_get_num:
+    .desc: Get the number of categories
+{
+    *num_cat = utarray_len(cat_table);
+}
+
+MPI_T_category_get_pvars:
+    .desc: Get performance variables in a category
+{ -- error_check -- indices
+    if (len != 0) {
+        MPIT_ERRTEST_ARGNULL(indices);
+    }
+}
+{ -- early_return --
+    if (len == 0) {
+        goto fn_exit;
+    }
+}
+
+MPI_T_cvar_get_index:
+    .desc: Get the index of a control variable
+{
+    name2index_hash_t *hash_entry;
+
+    /* Do hash lookup by the name */
+    HASH_FIND_STR(cvar_hash, name, hash_entry);
+    if (hash_entry != NULL) {
+        *cvar_index = hash_entry->idx;
+    } else {
+        mpi_errno = MPI_T_ERR_INVALID_NAME;
+        goto fn_fail;
+    }
+}
+
+MPI_T_cvar_get_info:
+    .desc: Get the information about a control variable
+    .skip: validate-ANY
+    .error: MPI_T_ERR_INVALID_INDEX
+{ -- error_check --
+    MPIT_ERRTEST_CVAR_INDEX(cvar_index);
+}
+{
+    cvar_table_entry_t *cvar;
+    cvar = (cvar_table_entry_t *) utarray_eltptr(cvar_table, cvar_index);
+
+    MPIR_T_strncpy(name, cvar->name, name_len);
+    MPIR_T_strncpy(desc, cvar->desc, desc_len);
+
+    if (verbosity != NULL)
+        *verbosity = cvar->verbosity;
+
+    if (datatype != NULL)
+        *datatype = cvar->datatype;
+
+    if (enumtype != NULL)
+        *enumtype = cvar->enumtype;
+
+    if (bind!= NULL)
+        *bind= cvar->bind;
+
+    if (scope != NULL)
+        *scope = cvar->scope;
+}
+
+MPI_T_cvar_get_num:
+    .desc: Get the number of control variables
+{
+    *num_cvar = utarray_len(cvar_table);
+}
+
+MPI_T_cvar_handle_alloc:
+    .desc: Allocate a handle for a control variable
+    .skip: validate-TOOL_MPI_OBJ
+
+MPI_T_cvar_handle_free:
+    .desc: Free an existing handle for a control variable
+{
+    MPIR_T_cvar_handle_t *hnd;
+    hnd = *handle;
+    MPL_free(hnd);
+    *handle = MPI_T_CVAR_HANDLE_NULL;
+}
+
+MPI_T_cvar_read:
+    .desc: Read the value of a control variable
+
+MPI_T_cvar_write:
+    .desc: Write a control variable
```

Comment:
missing errors `MPI_T_ERR_CVAR_SET_NOT_NOW` and `MPI_T_ERR_CVAR_SET_NEVER`

Reply:
I'll add a commit move the code in `mpit_impl.c: MPIR_T_cvar_write_impl` to `error_check` block and add the error code then.

Reply:
Added commit to address it.